bundles/org.openhab.binding.rotel/src/main/java/org/openhab/binding/rotel/internal/protocol/ascii/RotelAsciiV2ProtocolHandler.java

```diff
@@ -116,6 +124,44 @@ public byte[] buildCommandMessage(RotelCommand cmd, @Nullable Integer value) thr
         return message;
     }
 
+    @Override
+    public void handleIncomingData(byte[] inDataBuffer, int length) {
+        for (int i = 0; i < length; i++) {
+            boolean end = false;
+            if (searchKey && inDataBuffer[i] == '=') {
+                // End of key reading, check if the value is a fixed or variable length
+                searchKey = false;
+                byte[] dataKey = getDataBuffer();
+                String key = new String(dataKey, 0, dataKey.length, StandardCharsets.US_ASCII).trim();
+                variableLength = KEYSET.contains(key);
+                logger.trace("handleIncomingData: key = *{}* {}", key, variableLength ? "variable" : "fixed");
+                fillDataBuffer(inDataBuffer[i]);
+            } else if (searchKey) {
+                // Reading key
+                fillDataBuffer(inDataBuffer[i]);
+            } else if (inDataBuffer[i] == CHAR_END_RESPONSE) {
+                end = !variableLength || prevIsEndCharacter;
+            } else {
+                if (prevIsEndCharacter) {
+                    // End character inside a variable length value
+                    fillDataBuffer((byte) CHAR_END_RESPONSE);
+                }
+                // Reading value
+                fillDataBuffer(inDataBuffer[i]);
+            }
+            if (end) {
+                // End of value reading
+                handleIncomingMessage(getDataBuffer());
+                resetDataBuffer();
+                searchKey = true;
+                variableLength = false;
+                prevIsEndCharacter = false;
+            } else {
+                prevIsEndCharacter = inDataBuffer[i] == CHAR_END_RESPONSE;
+            }
+        }
+    }
```

Comment:
As far as I can understand from briefly looking at this method, it supports "variable length" in addition to the V1 parser. Would it be feasible to support the V1 algorithm by providing some option? Then the two parsers could be merged and redundancy eliminated.

Reply:
In fact, the variable length implementation is totally different in the two versions.
In V1, this will look like: key=7,my text
In V2, this will look like! key=my text$$
For fixed length, the V1 syntax is: key=value!
and the V2 syntax is: key=value$